### PR TITLE
Added eslint plugin to disallow using "import" keyword in debug comments

### PR DESCRIPTION
**Pull request by martnpaneq**
Internal (eslint-plugin-ckeditor5-rules): Added eslint plugin to disallow using the "import" keyword in debug comments. Closes ckeditor/ckeditor5#12479.

### CODE REVIEW

File: packages/eslint-plugin-ckeditor5-rules/lib/rules/use-require-for-debug-mode-imports.js [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+ * For licensing, see LICENSE.md.
+ */
+
+'use strict';
+
+module.exports = {
+	meta: {
+		type: 'problem',
+		docs: {
+			description: 'Disallow using import for functions used in debug mode.',
```
Comment (pomek) [inline packages/eslint-plugin-ckeditor5-rules/lib/rules/use-require-for-debug-mode-imports.js:12]:
```suggestion
			description: 'Disallow using the `import` keyword for modules used in debug mode.',
```

File: packages/eslint-plugin-ckeditor5-rules/lib/rules/use-require-for-debug-mode-imports.js [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+			category: 'CKEditor5'
+		},
+		fixable: 'code',
+		messages: {
+			'usingImportNotAllowed': 'Using import with "@if CK_DEBUG" keyword is not allowed. Please Use require().'
```
Comment (pomek) [inline packages/eslint-plugin-ckeditor5-rules/lib/rules/use-require-for-debug-mode-imports.js:17]:
```suggestion
			'usingImportNotAllowed': 'Using import with "@if CK_DEBUG_*" keyword is not allowed. Use `require()` instead.'
```

File: packages/eslint-plugin-ckeditor5-rules/lib/rules/use-require-for-debug-mode-imports.js [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+					} );
+				}
+			}
+		};
+	}
+};
+
+/**
+ * @param {String} [str='']
+ * @return {Boolean}
```
Comment (pomek) [inline packages/eslint-plugin-ckeditor5-rules/lib/rules/use-require-for-debug-mode-imports.js:52]:
```suggestion
 * @returns {Boolean}
```

We use `@returns` everywhere with [a single exception](https://github.com/ckeditor/ckeditor5-linters-config/blob/a3d8bb5aff7918094442d716dd782abe4b3260e0/packages/eslint-plugin-ckeditor5-rules/lib/rules/no-cross-package-imports.js#L54) that could be aligned too.

File: packages/eslint-plugin-ckeditor5-rules/lib/rules/use-require-for-debug-mode-imports.js [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+					}
+
+					context.report( {
+						node: comment,
+						messageId: 'usingImportNotAllowed',
+						fix( fixer ) {
+							const commentPrefix = comment.type === 'Line' ? '//' : '/*';
+							const commentSuffix = comment.type === 'Line' ? '' : '*/';
+							const regex = /import (.*) from (.*)?;/;
+							const replaceVal = 'const $1 = require($2);';
```
Comment (pomek) [inline packages/eslint-plugin-ckeditor5-rules/lib/rules/use-require-for-debug-mode-imports.js:39]:
```suggestion
							const replaceVal = 'const $1 = require( $2 );';
```
To align CKEditor 5 code style, let's add spaces around a package name (`$2`).

File: packages/eslint-plugin-ckeditor5-rules/lib/rules/use-require-for-debug-mode-imports.js [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+							const newCommentText = commentPrefix + comment.value.replace( regex, replaceVal ) + commentSuffix;
+							return fixer.replaceText( comment, newCommentText );
```
Comment (pomek) [inline packages/eslint-plugin-ckeditor5-rules/lib/rules/use-require-for-debug-mode-imports.js:41]:
```suggestion
							const newCommentText = commentPrefix + comment.value.replace( regex, replaceVal ) + commentSuffix;

							return fixer.replaceText( comment, newCommentText );
```

Missing a blank line before the return statement.

File: packages/eslint-plugin-ckeditor5-rules/lib/rules/use-require-for-debug-mode-imports.js [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+						continue;
+					}
+
+					context.report( {
+						node: comment,
+						messageId: 'usingImportNotAllowed',
+						fix( fixer ) {
+							const commentPrefix = comment.type === 'Line' ? '//' : '/*';
+							const commentSuffix = comment.type === 'Line' ? '' : '*/';
+							const regex = /import (.*) from (.*)?;/;
```
Comment (pomek) [inline packages/eslint-plugin-ckeditor5-rules/lib/rules/use-require-for-debug-mode-imports.js:38]: 1. The regular expression could be extracted as a constant value before `module.exports`.
2. Please, describe in a few words the purpose, e.g., `MODULE_IMPORT_REGEXP`.

File: packages/eslint-plugin-ckeditor5-rules/tests/use-require-for-debug-mode-imports.js [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+			{
+				code: '// @if CK_DEBUG // const { testFunction } = require(\'test-package\');'
+			},
+			{
+				code: '/**\n' +
+				'* @if CK_DEBUG // const TestPackage = require(\'test-package\');\n' +
+				'*/'
+			}
+		]
+	}
```
Comment (pomek) [inline packages/eslint-plugin-ckeditor5-rules/tests/use-require-for-debug-mode-imports.js:50]: I would see more tests as developers to ensure the rule works as expected.

1. `CK_DEBUG` is a generic flag. The testing environment supports also _named_ flags:
	* [`CK_DEBUG_MINIMAP`](https://github.com/ckeditor/ckeditor5/blob/30286f77b39526fce2856b03b9be0ba4cc91d1c7/packages/ckeditor5-minimap/src/minimap.js#L22)
	* [`CK_DEBUG_ENGINE`](https://github.com/ckeditor/ckeditor5/blob/a1f8243e0bdfedf3e71d91e64c390c8f3873086f/packages/ckeditor5-engine/src/model/operation/attributeoperation.ts#L207)
2. I quote [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import) - there are many import statements we should check, and if possibly, try to fix.
	* **An import of a default export module.**
		Invalid: 
		```js
		import defaultExport from "module-name";
		```
		Valid: 
		```js
		const defaultExport = require( "module-name" ).default;
		```
	* **An import of everything with a prefixed scope.**
		Invalid: 
		```js
		import * as name from "module-name";
		```
		Valid: 
		```js
		const name = require( "module-name" );
		```
	* **An import of a single, non-default module.**
		Invalid: 
		```js
		import { export1 } from "module-name";
		```
		Valid: 
		```js
		const { export1 } = require( "module-name" );
		```
		⚠️ We should also check if the import contains more modules. 
		All of them should be listed as `const { ... }`.
	* **An import of a default export module using brackets.**
		Invalid: 
		```js
		import { default as alias } from "module-name";
		```
		Valid: 
		```js
		const defaultExport = require( "module-name" ).default;
		```
	* **An import of a module under a changed name (using `as`).**
		Invalid: 
		```js
		import { export1 as alias1 } from "module-name";
		```
		Valid: 
		```js
		const { export1: alias1 ) = require( "module-name" );
		```

	* **Edge case: an import of non-module, e.g., a specific CSS loaded only for debugging purposes.**
		Invalid: 
		```js
		import "module-name";
		```
		Valid: 
		```js
		require( "module-name" );
		```
		I am not sure what is the expected statement. It needs to be verified manually.


The last question I have is – is the rule working with TypeScript files? I know that `import type ...` statements might fail, but we can safely ignore it as `// eslint-disable-line unused-var` will handle the problem. Types are not included in the output JavaScript files.

**Review by pomek — CHANGES_REQUESTED** [review]
At first look, everything looks good. However, we need to polish the new rule and support more options that the CKEditor 5 testing environment understands.

File: packages/eslint-plugin-ckeditor5-rules/lib/rules/use-require-for-debug-mode-imports.js [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+		};
+	}
+};
+
+/**
+ * @param {String} [str='']
+ * @returns {Boolean}
+ */
+function debugCommentDoesNotContainImport( str = '' ) {
+	return !/@if CK_DEBUG.* \/\/ import/.test( str );
```
Comment (psmyrek) [inline packages/eslint-plugin-ckeditor5-rules/lib/rules/use-require-for-debug-mode-imports.js:66]: Edge case: multiple debug imports could be aligned using white space 👇🏻 

```
// @if CK_DEBUG //     import ...
// @if CK_DEBUG_FOO // import ...
// @if CK_DEBUG_X //   import ...
```

```suggestion
	return !/@if CK_DEBUG.* \/\/\s*import/.test( str );
```

File: packages/eslint-plugin-ckeditor5-rules/lib/rules/use-require-for-debug-mode-imports.js [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+ */
+
+'use strict';
+
+const DEFAULT_ALIAS_IMPORT = /import { default as (.*) } from (.*)?;/;
+const NAMED_ALIAS_IMPORT = /import { (.*) as (.*) } from (.*)?;/;
+const NAMED_IMPORT = /import { (.*) } from (.*)?;/;
+const NAMESPACE_ALIAS_IMPORT = /import .* as (.*) from (.*)?;/;
+const DEFAULT_IMPORT = /import (.*) from (.*)?;/;
+const SIDE_EFFECT_IMPORT = /import (.*)?;/;
```
Comment (psmyrek) [inline packages/eslint-plugin-ckeditor5-rules/lib/rules/use-require-for-debug-mode-imports.js:13]: Currently, the following debug import is not fixed as we expect, because only the first import is replaced but the second one is not:

```
/**
 * @if CK_DEBUG // import defaultExport from 'module-name';
 * @if CK_DEBUG // import defaultExport2 from 'module-name2';
 */
```

```suggestion
const DEFAULT_ALIAS_IMPORT = /import { default as (.*) } from (.*)?;/g;
const NAMED_ALIAS_IMPORT = /import { (.*) as (.*) } from (.*)?;/g;
const NAMED_IMPORT = /import { (.*) } from (.*)?;/g;
const NAMESPACE_ALIAS_IMPORT = /import .* as (.*) from (.*)?;/g;
const DEFAULT_IMPORT = /import (.*) from (.*)?;/g;
const SIDE_EFFECT_IMPORT = /import (.*)?;/g;
```

Comment (pomek) [reply packages/eslint-plugin-ckeditor5-rules/lib/rules/use-require-for-debug-mode-imports.js:13]: I do not think this expression is correct at all. See: https://github.com/ckeditor/ckeditor5/blob/a1f8243e0bdfedf3e71d91e64c390c8f3873086f/packages/ckeditor5-engine/src/model/operation/attributeoperation.ts#L207-L211.

**Review by pomek — APPROVED** [review]
LGTM.
